src/libraries/System.Memory.Data/src/System/BinaryData.cs

```diff
@@ -330,6 +330,82 @@ private static async Task<BinaryData> FromStreamAsync(Stream stream, bool async,
+            }
+
+            return FromFileAsync(path, async: false).GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified file
+        /// and sets <see cref="MediaType"/> to <see pref="mediaType"/> value.
+        /// </summary>
+        /// <param name="path">The path to the file.</param>
+        /// <param name="mediaType">MIME type of this data, e.g. <see cref="MediaTypeNames.Application.Octet"/>.</param>
+        /// <returns>A value representing all of the data from the file.</returns>
+        /// <seealso cref="MediaTypeNames"/>
+        public static BinaryData FromFile(string path, string? mediaType)
+        {
+            if (path is null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            return FromFileAsync(path, async: false, mediaType).GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified file.
+        /// </summary>
+        /// <param name="path">The path to the file.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data from the file.</returns>
+        public static Task<BinaryData> FromFileAsync(string path, CancellationToken cancellationToken = default)
+        {
+            if (path is null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            return FromFileAsync(path, async: true, cancellationToken: cancellationToken);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified file
+        /// and sets <see cref="MediaType"/> to <see pref="mediaType"/> value.
+        /// </summary>
+        /// <param name="path">The path to the file.</param>
+        /// <param name="mediaType">MIME type of this data, e.g. <see cref="MediaTypeNames.Application.Octet"/>.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data from the file.</returns>
+        /// <seealso cref="MediaTypeNames"/>
+        public static Task<BinaryData> FromFileAsync(string path, string? mediaType,
+            CancellationToken cancellationToken = default)
+        {
+            if (path is null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            return FromFileAsync(path, async: true, mediaType, cancellationToken);
+        }
+
+        private static async Task<BinaryData> FromFileAsync(string path, bool async,
+            string? mediaType = default, CancellationToken cancellationToken = default)
+        {
+            using FileStream fileStream = File.OpenRead(path);
```

Comment:
Would it be better to use `File.ReadAllBytes`? Since it has that tries to minimize allocations of the byte array if the file-size is determinable.

Reply:
Good question!
- The file-size optimization is also supported in the `FromStream` flow.

- The `ReadAllBytesAsync` method is not available for the net462 and netstandard2.0 TFMs. Can we then fallback to the synchronous implementation for these TFMs?

- The ReadAllBytes{Async} flow may be more efficient to avoid the intermediate copy buffer of the CopyTo{Async} flow.

Here would be the corresponding code:
```cs
        private static
#if NET6_0_OR_GREATER
        async
#endif
        Task<BinaryData> FromFileAsync(string path, bool async, string? mediaType = default, CancellationToken cancellationToken = default)
        {
#if NET6_0_OR_GREATER
            byte[] bytes;

            if (async)
            {
                bytes = await File.ReadAllBytesAsync(path, cancellationToken).ConfigureAwait(false);
            }
            else
            {
                bytes = File.ReadAllBytes(path);
            }

            return new BinaryData(bytes, mediaType);
#else
            byte[] bytes = File.ReadAllBytes(path);

            return Task.FromResult(new BinaryData(bytes, mediaType));
#endif
        }
```

Better?

Reply:
Ah I see. Thanks for looking into it. Personally I don't think the potential benefit outweighs the added complexity, I would keep it as it is then.